crates/router/src/connector/dlocal/transformers.rs

```diff
@@ -0,0 +1,588 @@
+use std::{collections::HashMap, ptr::eq};
+
+use error_stack::{IntoReport, ResultExt};
+use masking::PeekInterface;
+use reqwest::Url;
+use serde::{Deserialize, Serialize};
+use storage_models::schema::payment_attempt::payment_method_id;
+
+use crate::{
+    core::errors,
+    services,
+    types::{self, api, storage::enums},
+};
+
+#[derive(Debug, Default, Eq, PartialEq, Serialize)]
+pub struct Payer {
+    pub name: String,
+    pub email: String,
+    pub document: String,
+}
+
+#[derive(Debug, Default, Eq, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Card {
+    pub holder_name: String,
+    pub number: String,
+    pub cvv: String,
+    pub expiration_month: i32,
+    pub expiration_year: i32,
+    pub capture: String,
+    pub installments_id: Option<String>,
+    pub installments: Option<String>,
+}
+
+#[derive(Debug, Default, Eq, PartialEq, Serialize)]
+pub struct ThreeDSecureReqData {
+    pub force: bool,
+}
+
+//TODO: Fill the struct with respective fields
+#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+pub struct DlocalPaymentsRequest {
+    pub amount: i64, //amount in cents, hence passed as integer
+    pub currency: enums::Currency,
+    pub country: Option<String>,
+    pub payment_method_id: String,
+    pub payment_method_flow: String,
+    pub payer: Payer,
+    pub card: Option<Card>,
+    pub order_id: String,
+    pub notification_url: String,
+    pub three_dsecure: Option<ThreeDSecureReqData>,
+    pub callback_url: Option<String>,
+    // pub wallet: Option<String>,
+}
+
+impl TryFrom<&types::PaymentsAuthorizeRouterData> for DlocalPaymentsRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(item: &types::PaymentsAuthorizeRouterData) -> Result<Self, Self::Error> {
+        match item.request.payment_method_data {
+            api::PaymentMethod::Card(ref ccard) => {
+                let should_capture = matches!(
+                    item.request.capture_method,
+                    Some(enums::CaptureMethod::Automatic) | None
+                );
+                let payment_request = Self {
+                    amount: item.request.amount,
+                    country: Some(get_currency(item.request.currency)),
```

Comment:
If country is mandatory in the request get it from billing address in the `PaymentsAuthorizeRouterData`. Remove get_currency method.

Reply:
dlocal has a check of country and currency code, now if we take country from billing address, it might not be correct for all the cases.

Reply:
If the payments fail because of miss match between country and currency, let it fail. We should not alter/add any data in the payment create request. Merchant should ensure right billing country and currency.